### ISSUE TITLE
Mixing Callbacks & Promises in native code

### ISSUE THREAD

**Issue by Ehesp** [facebook/react-native#14702]
### Description

On Android (need to check iOS) there seems to be an issue (don't see any documentation regarding this) on the native side, whereby I'm unable to mix callbacks and promises. If my method returns a promise, any callbacks passed through before are `null`. 

### Reproduction Steps

JS side:

```js
NativeModule
  .someMethod('name', () => {

  })
  .then()
  .catch();

```

Android side:

```java
@ReactMethod
public void someMethod(String name, final Callback callback, final Promise promise) {
  // name = "name"
  // callback = null
  // promise = Promise
}
```

### Additional Information

* React Native version: 45.1
* Platform: Android
* Development Operating System: Mac
* Build tools: Android Studio 3 - Gradle 4

**Comment by brunolemos** [comment]
AFAIK react native accept them as last argument only.

From the [docs](https://facebook.github.io/react-native/docs/native-modules-ios.html#promises): 
> Promises
When the last parameters of a bridged native method are an RCTPromiseResolveBlock and RCTPromiseRejectBlock, its corresponding JS method will return a JS Promise object.

For callback is probably the same behavior: works when it is the last argument.

**Comment by ide** [comment]
You can't implement a promise and callbacks together in this context. React Native's bridge is designed to support up to two callbacks -- one for successful results and one for errors. The resolver and rejecter functions of promises are implemented using these two callbacks.

AFAIK there isn't a fundamental reason why the callbacks are still supported. I believe we should work towards deprecating them (over several months, there's no need for it to be a sudden breaking change). Native promises establish a better mental model of asynchronously returning a single result or error, which matches how the React Native bridge works.

tl;dr: use only promises when writing native modules

**Comment by Ehesp** [comment]
Thanks for the reply @ide. That makes sense, however it'll make the implementation for one Firebase SDK methods rather tricky. 

I guess I'll have to use an event emitter and handle it JS side, don't see any other way around it! 

Cheers

**Comment by ide** [comment]
An event emitter is the correct way to push information from native to JS (in contrast with returning a single value that JS asks to pull from native).
